### PR TITLE
Blocks: add a new Cookie Consent Block

### PR DESCRIPTION
**Pull request by jeherve**
## Proposed changes:

Replay of #29191 in the Jetpack plugin.

This adds a new Beta block, to be used in block themes, to display a cookie consent banner on the site.

image

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion

* pdDR7T-Ax-p2
* pdDR7T-Cr-p2

## Does this pull request change what data or activity we track or use?

* Yes, it adds a new cookie (`eucookielaw`) when you click to accept the banner. That cookie is already in use by the Cookies & Consents Banner widget, but that should still be documented.

## Testing instructions:

1. Start with a site that's connected to WordPress.com and where you have enabled Beta blocks (`add_filter( 'jetpack_blocks_variation', function () { return 'beta'; } );`).
    * On WordPress.com Simple, you do not need the filter. You will see the block when you are logged in.
2. Go to Appearance > Themes and enable a block theme; Twenty Twenty Three will work.
3. Go to Appearance > Site Editor
4. You should now be able to add and customize the block. 
    * Check that you can edit contents and change the settings in the block sidebar.
    * Check that your changes can be saved.
    * Check that your changes are visible on the frontend.
    * When dismissing the banner in the frontend, you will see the `eucookielaw` cookie be added in your browser. The banner should then be gone.
    * When refreshing the page, the banner should still be gone.

### CODE REVIEW

File: projects/plugins/jetpack/extensions/blocks/cookie-consent/cookie-consent.php
Comment (jeherve) [inline projects/plugins/jetpack/extensions/blocks/cookie-consent/cookie-consent.php]: Will this work on self-hosted sites and WoA sites? That `permanent_noadverts` option is not available on self-hosted sites, for example.

File: projects/plugins/jetpack/extensions/blocks/cookie-consent/cookie-consent.php
Comment (jeherve) [inline projects/plugins/jetpack/extensions/blocks/cookie-consent/cookie-consent.php]: This is there for development, it should be removed before we merge this PR.

File: projects/plugins/jetpack/extensions/blocks/cookie-consent/icon.js
Comment (jeherve) [inline projects/plugins/jetpack/extensions/blocks/cookie-consent/icon.js:1]: This is a default icon. We can use a custom icon instead (or remove it and use the shield Gridicon instead like you did in #29191).

File: projects/plugins/jetpack/extensions/blocks/cookie-consent/index.js
Comment (jeherve) [inline projects/plugins/jetpack/extensions/blocks/cookie-consent/index.js]: We'd need to address that.

File: projects/plugins/jetpack/extensions/blocks/cookie-consent/cookie-consent.php
Comment (alshakero) [reply projects/plugins/jetpack/extensions/blocks/cookie-consent/cookie-consent.php]: ~~I think that should be OK. Because it would default to `false`, which is what we want.~~

Oh man, I forgot we flip the entire condition.

File: projects/plugins/jetpack/extensions/blocks/cookie-consent/attributes.js
Comment (anomiex) [inline projects/plugins/jetpack/extensions/blocks/cookie-consent/attributes.js:9]: I suspect that self-hosted sites would need a different link here. Maybe Atomic too?

File: projects/plugins/jetpack/extensions/blocks/cookie-consent/index.js
Comment (anomiex) [inline projects/plugins/jetpack/extensions/blocks/cookie-consent/index.js]: This will add another instance of #26792 that would need fixing soon.

File: projects/plugins/jetpack/extensions/blocks/cookie-consent/attributes.js
Comment (jeherve) [reply projects/plugins/jetpack/extensions/blocks/cookie-consent/attributes.js:9]: Yes, for the widget we allow custom links in there, and also take into account the page you can set as your privacy policy page in WordPress:
https://github.com/Automattic/jetpack/blob/3104e377ae2d45ad2d4ffadfd839cee47aa871df/projects/plugins/jetpack/modules/widgets/class-jetpack-eu-cookie-law-widget.php#L138-L139

This is something we could improve on in a follow-up PR I think.

File: projects/plugins/jetpack/extensions/blocks/cookie-consent/index.js
Comment (jeherve) [reply projects/plugins/jetpack/extensions/blocks/cookie-consent/index.js]: Good call. Given our timeline (WordPress 6.2 landing in March), that's probably something we want to fix in here before we merge this.

File: projects/plugins/jetpack/extensions/blocks/cookie-consent/index.js
Comment (alshakero) [reply projects/plugins/jetpack/extensions/blocks/cookie-consent/index.js]: Do we have to add anything here if the block as a default value for every attribute?

File: projects/plugins/jetpack/extensions/blocks/cookie-consent/attributes.js
Comment (alshakero) [reply projects/plugins/jetpack/extensions/blocks/cookie-consent/attributes.js:9]: It would be a nice improvement for the default value, but the block is entirely customizable and users can change the URL.

**Comment by alshakero** [comment]
Thank you so much @jeherve and @anomiex! I addressed all your feedback. 

- Added a nice MIT licensed icon.
- Removed the exclusion for sites with ads, took inspiration [from the widget](https://github.com/Automattic/jetpack/blob/trunk/projects/plugins/jetpack/modules/widgets/class-jetpack-eu-cookie-law-widget.php#L335). I took this condition from `wpcom/wp-content/mu-plugins/widgets/eu-cookie-law/eu-cookie-law.php` but it doesn't seem to translate well in Jetpack (at least locally). And removing it is fairly harmless since the block is opt-in.
- Made the description plain text. Hence, removed the learn more link.

File: projects/plugins/jetpack/extensions/blocks/cookie-consent/edit.js
Comment (jeherve) [inline projects/plugins/jetpack/extensions/blocks/cookie-consent/edit.js:95]: It would be nice if I had more flexibility over the button, such as alignment within the block. Maybe we could give more options by not locking the button? 

Since the button has to be there for the block to work, we could maybe take an approach similar to #24838?

File: projects/plugins/jetpack/extensions/blocks/cookie-consent/edit.js
Comment (jeherve) [inline projects/plugins/jetpack/extensions/blocks/cookie-consent/edit.js:89]: I wonder if we would get more options by switching to using `jetpack/button`, like we do in most of our other blocks that use buttons?

File: projects/plugins/jetpack/extensions/blocks/cookie-consent/icon.js
Comment (jeherve) [inline projects/plugins/jetpack/extensions/blocks/cookie-consent/icon.js:7]: I like that new icon, it looks perfect!

File: projects/plugins/jetpack/extensions/blocks/cookie-consent/attributes.js
Comment (jeherve) [inline projects/plugins/jetpack/extensions/blocks/cookie-consent/attributes.js:19]: Something seems off with this. It grabs the colors just fine, but when I go to edit them, I see this:

image

File: projects/plugins/jetpack/extensions/blocks/cookie-consent/index.js
Comment (jeherve) [inline projects/plugins/jetpack/extensions/blocks/cookie-consent/index.js]: Maybe there could be a short sentence describing what this is for, and then this sentence explaining how to use it?

File: projects/plugins/jetpack/extensions/blocks/cookie-consent/edit.js
Comment (jeherve) [inline projects/plugins/jetpack/extensions/blocks/cookie-consent/edit.js:58]: Should we offer a center option here as well?

File: projects/plugins/jetpack/extensions/blocks/cookie-consent/edit.js
Comment (jeherve) [inline projects/plugins/jetpack/extensions/blocks/cookie-consent/edit.js:86]: Could we allow headings here as well, for folks who want to add a bigger title before the explanation?

File: projects/plugins/jetpack/extensions/blocks/cookie-consent/edit.js
Comment (jeherve) [inline projects/plugins/jetpack/extensions/blocks/cookie-consent/edit.js]: Maybe this could simply be "Block Settings"?

**Review by jeherve — COMMENTED** [review]
I gave it a try. I think this first iteration is ready to merge; we can iterate in follow-up PRs.

I cannot approve it though since I opened the PR, so I'll let someone else from Crew take care of that.

I left some more comments below.


*******

I wonder if some folks will be tempted to do something like this:

image

And if so, should we support having 2 buttons, one accept, one reject? That's probably something for a v2.

File: projects/plugins/jetpack/extensions/blocks/cookie-consent/edit.js
Comment (alshakero) [reply projects/plugins/jetpack/extensions/blocks/cookie-consent/edit.js:95]: We actually started with a much more flexible block, we had the whole thing as a template (see: pdDR7T-Cr-p2) but the flexibility makes it a great tool to abuse and show one-time ads on your site.

The only reason I see `InnerBlocks` here is that I couldn't figure hard-coding a customizable button. `RichText` only allowed us to hard-code a `p` but I was unable to find a way for a button.

File: projects/plugins/jetpack/extensions/blocks/cookie-consent/attributes.js
Comment (alshakero) [reply projects/plugins/jetpack/extensions/blocks/cookie-consent/attributes.js:19]: I couldn't come up with a way to make this work. Undocumented and I tried everything I can come up with!

File: projects/plugins/jetpack/extensions/blocks/cookie-consent/edit.js
Comment (alshakero) [reply projects/plugins/jetpack/extensions/blocks/cookie-consent/edit.js:58]: It took us so much work to get this running in more than one theme.

File: projects/plugins/jetpack/extensions/blocks/cookie-consent/edit.js
Comment (alshakero) [reply projects/plugins/jetpack/extensions/blocks/cookie-consent/edit.js:86]: These `InnerBlocks` are only around the button. They're not meant to give control to the user. They're just a way to make the button customizable. I wish I could lock it down further.

**Review by alshakero — APPROVED** [review]
Pushed some improvements and it looks great now!

**Review by anomiex — APPROVED** [review]
I'll go with Jeremy saying it's good enough to merge and iterate on.

**Pull request merged**
